JP-3364 allow wcs shifts for IFU individual exposures in cube_build

jemorrison opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3364](https://jira.stsci.edu/browse/JP-3364)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #7876

<!-- describe the changes comprising this PR here -->
This PR addresses allows the user to provide a list of ra and decs for each input file. The shifts are provided in arc seconds 
prior to the projections into the xi/eta space (tangent plane). This allows for a cube-space fix to the known WCS issue based on user-defined inputs.  For each input file in the association cube_build creates the cubes there must be a ra and dec shift provided in an OFFSET file.


An ifuoffset.schema.yaml file was added to the cube_build step (instead of stdatamodels). This schema file defines the format of the offset file. 
 The offsets are defined by a filename, raoffset, and decoffset list. The units of the offset are required to be arcsec and are required to be in the offset file. This offset file must be an asdf file.  
 An example of how to make the offset files is provided in the documentation. 

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1723/
- [x] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

jemorrison commented:
Example of how to make an offset file:
It is assumed the user will determine the ra and dec offsets for each file
filename = list of filenames that are the same as in the association
raoffset = list of ra shifts cooresponding to each filename 
decoffset = list of dec shifts  cooresponding to each filename 
import asdf
offsets = {}

offsets['filename'] = []
offsets['raoffset'] = []
offsets['decoffset'] = []
for i in range(num):
    offsets['filename'].append(filename[I])
    offsets['raoffset'].append(ra_center1[i])
    offsets['decoffset'].append(dec_center1[i])
    
af = asdf.AsdfFile({'offsets':offsets})
af.write_to(input_dir  + 'offsets.asdf')

karllark reviewed (changes requested):
The code looks good to me.

I did notice there was no update to the docs.  This is important as the format of the file and the units of the offsets are needed info for users.  These were questions I had that I found by reviewing the code changes.  Updating the docs with this new option should be done.

jemorrison commented:
@karllark I will definitely update the docs. I just want to make sure the approach I have made is what is expected. I need to include docs on how to make the asdf offset file.

jemorrison commented:
@tapastro I have included in a comment the code I used to create the asdf offset file. Does look about right ?
The code is expected a certain format. I don't know of any way to make it general so I was thinking we could provide an example of how to make the ra,dec shifts needed in the offset file

karllark commented:
> @karllark I will definitely update the docs. I just want to make sure the approach I have made is what is expected. I need to include docs on how to make the asdf offset file.

Great.  Yes, the approach makes sense to me.

drlaw1558 commented:
There was a comment about how you made the asdf file?  I think I'm making mine incorrectly, as I'm getting errors using it.

drlaw1558 commented:
One thing I noticed- the code matches on local filenames, while associations can sometimes be full-path filenames.  That's probably ok, but worth noting in the documentation that the ASDF file should contain only the filenames and not the full paths otherwise it won't match them properly.

We'll also definitely need to give the above code to create input files explicitly in the documentation as ASDF files are very non-intuitive for most folks.

Another thought: we probably want the offsets to be used in actual arcseconds rather than difference in sky coordinates.  I.e., in the test case I'm using (PID 1523) I'm specifying an offset of 1.0 in both RA and in DEC, but the source is barely moving in the RA direction because of the cos(dec) term (this case is at 70 degrees South).  We could choose either convention, but I think having offsets be consistent units of arcsec regardless of where we are on the sky makes most sense, and handle the cos(dec) translation to sky coordinates under the hood.

jemorrison commented:
@drlaw1558 I was wondering about the cos(dec) term. Since the user is creating the offset file I was not sure what made the most sense in how to specific what the ra offset should be. I used IFU cubes created from a single file and then centroided on the source to figure out a set of ra and dec offsets than I then feed back to cube build. I assume the cos(dec) term we need to apply is the dec of CRVAL2. Is this too circular ?  Should we provide one work flow on how to do this. 1. Build a cube at a given crval1 and crval2 determine the ra and dec offset. This offset file is then provided to a new run of cube_build with the same crval1, crval2. Or am I thinking about this too narrowly

jemorrison commented:
I don't think I have the cos(dec) term correct. I am dividing by it - thinking when you are at large dec you want the offset to  larger in ra . But the more I think about this I think I should just have this reviewed and someone can point out my mistakes cause that does not feel right to me. I might need a clearer description in the docs on that ra and dec offset - how they are determined - absolute not relative.

jemorrison commented:
I can not get the readthedocs built. Locally I get failures I have not idea how to fix:
(Extension error (sphinx_automodapi.automodsumm):
Handler <function process_automodsumm_generation at 0x12c99e840> for event 'builder-inited' threw an exception (exception: No module named 'jwst.resample.resample_utils')

In the past I have used the cl results to look for doc error but I am not finding what is in the 'Details' at all helpful

melanieclarke commented:
> I can not get the readthedocs built.

I tested your branch locally in my environment, and was able to build them (with some warnings), so it's probably an issue with your setup/local repo rather than the changes to the docs.  Are you running `make clean` before `make html`?

melanieclarke commented:
Also, searching for 'warn' in the details from the docs build on CI, I think this is the relevant problem:

```
/home/docs/checkouts/readthedocs.org/user_builds/jwst-pipeline/checkouts/8720/docs/jwst/cube_build/arguments.rst:134: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/jwst-pipeline/checkouts/8720/docs/jwst/cube_build/arguments.rst:153: ERROR: Unexpected indentation.
```

melanieclarke commented:
And one more comment, about getting the cosine term right -- it might be better to use an astropy SkyCoord or similar to do the offset math for you.

melanieclarke commented:
> @melanieclarke what version of sphinx should I have. I have Sphinx v7.4.7

The Sphinx version shouldn't be terribly important, as long as it's <8.  Upgrading mine to 7.4.7 I am still able to build the docs on your branch.  

From the error message, it sounds like it's looking for a jwst import it can't find.  If you're able to run the pipeline okay from the environment you're making the docs in, then it may be an issue with some old documentation artifacts or something like that - you could try doing a `git clean` in the docs directory to remove.  

Otherwise, I'd recommend rebuilding your python environment and see if a fresh start sorts you out.

nden commented on jwst/cube_build/ifu_cube.py:
```diff
@@ -1809,6 +1878,21 @@ def map_nirspec_pixel_to_sky(self, input_model):
+            log.info("Ra and dec offset (arc seconds) applied to file :%5.2f, %5.2f,  %s",
+                     raoffset, decoffset, filename)
+            raoffset = raoffset/3600.0 # convert to degress
+            decoffset = decoffset/3600.0
```
Lines 1882 - 1894 are repeated three times. Might be better to have a function.

jemorrison replied:
pulled out modifying units to after finding median dec. So it is not just in one place.

jemorrison commented on jwst/cube_build/cube_build_step.py:
```diff
@@ -524,3 +539,48 @@ def read_user_input(self):
             self.pars_input['grating'] = list(set(self.pars_input['grating']))
 # ________________________________________________________________________________
+
+    def check_offset_list(self):
```
I asked for suggestion from the team and Brett suggested I do those checks. I can change it I am just not sure what you mean by let asdf do the validation. I might have also misunderstood what Brett was suggesting.

jemorrison commented:
@drlaw1558 @nden 
I have made a draft IFUOffsetModel and I am trying to build an offset file with it.  I have never done this before so I could use some pointers. I made it an ReferenceFileModel because I did not see another Class model I could use. 
 But that gives a bunch of warnings of not having meta data. 
I really did not want this to be difficult to the user to create so I set the values that it was complaining about 
pedigree, author, use after, description. I tried to also set 'retype' and 'instrument.name' - but it is not accepting my values 

Is this the right approach ? 
https://github.com/spacetelescope/stdatamodels/pull/322

drlaw1558 commented:
@jemorrison Is this IFUOffsetModel file just to provide a template for the user-input offsets?  In all honesty what I think would be simplest from a user point of view is a plaintext file with 3 columns: name, ra offset, dec offset.  No need for any special wrapping, just three columns of text.

jemorrison commented:
@drlaw1558  So are you suggesting not to make an asdf file ? That was my first approach but it was suggested to me that the preferred standard way to have "reference type files " in the pipeline was asdf. Which is why I went that route. I thought the easiest way was a dictionary so I have could have a "key" that marked what the data - raoffset or decoffset. 
I tried to make the validation in the cube_build code on the asdf file but I guess that is not standard and making a schema is.
So now I am trying to make a schema with a dictionary. Is the schema - over-kill.  The code I wrote in the documentation and in the ticket is all a user needs to do- they will not need to use the schema to create the file. The schema will just be used in cube_build to confirm the reference file is of the correct format which I guess is the standard way. I don't know suggestions are welcome.

drlaw1558 commented:
For point of comparison, what does the tweakreg input look like?  That's a conceptually very similar step, which also allows for a user-defined input file tweakreg_catalog.  In that case it lists source coordinates within an image rather than offsets between images, but similarities nonetheless in that it should be easy for users to create.

The more we can simplify things for users the better- either with a plaintext style input format, or (if we do use an asdf file) a convenience function that they can call in order to create the file with their desired input values.

jemorrison commented:
@nden I am trying something that was  discussed in standup and that is to put the schema file in cube_build rather than in stdatamodels. 
 I have created an ifuoffset.schema.yaml file and placed it in cube_build directory (you can see  in this pr) . But I do not know how to use it to validate the offset file. Do I need a wrapper python file around it ?
 If that is one way to do it - where would the  schema_url  point to. 
 
 Currently in the check_offset_file module I have this line
  af = asdf.open(self.offset_file, custom_schema = 'ifuoffset.schema.yaml')
and it can not find this yaml file. 
I tried to import it but I can not figure out to I can import a yaml file. Maybe  I need a python wrapper around it and import the python file

melanieclarke commented on jwst/cube_build/cube_build_step.py:
```diff
@@ -38,7 +39,7 @@ class CubeBuildStep (Step):
 
     spec = """
          channel = option('1','2','3','4','all',default='all') # Channel
-         band = option('short','medium','long','short-medium','short-long','medium-short', \
+         band = option('short','medium','long','short-medium','short-long','medium-short',
```
Testing on your branch now to look at the schema file.  This change is causing a parsing error - restoring the backslash at the end of the first line lets the step run.

jemorrison commented:
@melanieclarke I can work with this method to get my schema set up correctly but it requires that the scheme file be located where I am running cube_build (in a data directory). I wanted to store the schema file in the cube_build directory. 
So how do we point to the cube build directory in a general way ?

melanieclarke commented:
> @melanieclarke I can work with this method to get my schema set up correctly but it requires that the scheme file be located where I am running cube_build (in a data directory). I wanted to store the schema file in the cube_build directory. So how do we point to the cube build directory in a general way ?

The above suggestion does point to the schema in the cube_build directory.  In Python, `__file__` is the filename for the current file where the code is written (cube_build.py), so `Path(__file__).parent` is the cube build directory.  Did the above suggestion not work for you?

braingram commented:
One option is to define the schema in the code (by picking a uri and adding the uri and yaml as a "resource_mapping" for asdf). Here's an example:
```python
import asdf

uri = "asdf://stsci.edu/jwst/offsets.yaml"
asdf.get_config().add_resource_mapping({uri: b"""%YAML 1.1
---
type: object
properties:
    offsets:
        type: array
        items:
            type: object
            properties:
                filename:
                    type: string
                ra:
                    type: number
                dec:
                    type: number
"""})

offsets = [
    {
        "filename": "foo.fits",
        "ra": 0.0,
        "dec": 0.0,
    },
    {
        "filename": "bar.fits",
        "ra": 1.0,
        "dec": 2.0,
    },
]

af = asdf.AsdfFile({"offsets": offsets}, custom_schema=uri)

bad_offsets = [
    {
        "filename": "bam.fits",
        "ra": "42",
        "dec": 0.0,
    },
]
af = asdf.AsdfFile({"offsets": bad_offsets}, custom_schema=uri)
```

jemorrison commented:
Ok got it now. I did not 'import the pathlib import Path' correctly
Thanks

jemorrison commented:
@melanieclarke I am struggling with testing the code check_offset_file in cube_build_step.
I put what I was thinking in a test in test_configuration.py the last test: test_offset_file_config

I get an error in the test when it encounters self.log.error
Maybe I have set up check_offset_file incorrectly to try and catch ValidationErrors from the schema.
Suggestions ?

melanieclarke commented on jwst/cube_build/tests/test_configuration.py:
```diff
@@ -467,3 +511,35 @@ def test_calspec3_config_nirspec_multi(tmp_cwd, nirspec_medium_coverage):
+        par_filename,
+        **pars)
+
+    offsets = CubeBuildStep.check_offset_file(cubeinfo.input_models)
```
@jemorrison - this line is why you're seeing an error in the test.  Calling it this way sets `self` to `cubeinfo.input_models`.

You need to instantiate CubeBuildStep first, then you can use the instantiated object to call the `check_offset_file` function.

jemorrison commented:
@melanieclarke 
Should I move the check_offset_file out of cube_build_step.py class. I would like to test cases that it fails and also that it passes. But I don't see a way to call CubeBuildStep and only call the check_offset_file, is that possible. The self in check_offset_file is part of the issue. If I just past the input models, offset file to this routine it would be easier to test

melanieclarke commented:
> @melanieclarke Should I move the check_offset_file out of cube_build_step.py class. I would like to test cases that it fails and also that it passes. But I don't see a way to call CubeBuildStep and only call the check_offset_file, is that possible. The self in check_offset_file is part of the issue. If I just past the input models, offset file to this routine it would be easier to test

This should be possible to test without moving it out of the class -- you just need to call it differently.
Instead of this:
```
offsets = CubeBuildStep.check_offset_file(cubeinfo.input_models)
```
Try something like
```
step = CubeBuildStep()
step.input_models = cubeinfo.input_models
step.offset_file = offset_file
step.check_offset_file()
```

jemorrison commented:
@melanieclarke  I updated the change log and added it to this PR. But I keep getting a RED X on 'Ensure changelog' ...... exists for this pull . I don't think there is an error for the line I added. 

I am running the regression tests now for this branch

melanieclarke commented:
> I updated the change log and added it to this PR. But I keep getting a RED X on 'Ensure changelog' ...... exists for this pull . I don't think there is an error for the line I added.

It looks like you added the JP ticket number instead of the PR number - the check needs a GitHub cross-reference.  It should be `[#8720]`

jemorrison commented:
@melanieclarke Goodness that was stupid. Thanks

jemorrison commented:
@melanieclarke @nden @drlaw1558 Ready for final review.
I changed the way the offset file is created - it is now just 3 lists, but it is still as asdf file. 
I ran the regression tests and 1 failure that is not related to this PR.
I added a unit test but I did not add a regression test for this new option.
If that is desired let me know

melanieclarke commented on jwst/cube_build/ifu_cube.py:
```diff
@@ -1306,13 +1309,47 @@ def setup_ifucube_wcs(self):
+                decs = []
+                for k in range(n):
+                    input_file = self.master_table.FileMap[self.instrument][this_a][this_b][k]
+                    input_model = datamodels.open(input_file)
```
This datamodels.open call seems odd - is looks like that element of the FileMap ought to already be an open datamodel, not a filename?

melanieclarke commented on jwst/cube_build/ifu_cube.py:
```diff
@@ -1306,13 +1309,47 @@ def setup_ifucube_wcs(self):
+                self.median_dec = np.nanmedian(decs)
+                # fold in the median_dec information into ra offset
+                noffsets = len(self.offsets['raoffset'])
+                # convert ra and dec offsets to degrees and adjust ra offset for cos(dec) 
+                for im in range(noffsets):
+                    self.offsets['raoffset'][im] = (self.offsets['raoffset'][im]/3600.0)/np.cos(self.median_dec*deg2rad)
+                    self.offsets['decoffset'][im] = self.offsets['decoffset'][im]/3600.0
```
Did you look into using SkyCoord to do the sky math instead?  This median dec handling is probably fine in most cases, but I always worry about edge cases with spherical coordinates.

jemorrison replied:
I did but we wanted the cos(dec) to be the same for all the time. We did not want that to vary.

melanieclarke replied:
Sorry, I wasn't clear here.  Using one declination reference point is fine, my concern is that sky math has a lot of edge cases and gotchas which using something like SkyCoord naturally works around.  Doing the offset calculations by hand seems easier than setting up SkyCoord objects, but it might be more error prone.

Right now, I can't think of a counterexample where taking a simple median over declination values would be a problem, but it did occur to me over the weekend that later, when you're adding the offset to the RA and Dec, there is no wraparound check.  What happens if the declination is near 90 and the offset pushes it over?  What if the RA is near 0 or 360?  We can add checks for those cases everywhere the addition happens, or we could refactor with SkyCoord and let it do the math.

melanieclarke replied:
That's not quite what I was suggesting.  What I meant by using SkyCoords was to make the corner values that eventually use the offsets SkyCoord objects, and then use something like [SkyCoord.directional_offset_by](https://docs.astropy.org/en/latest/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.directional_offset_by) to compute the new RA/Dec position from the input offset in arcsec.  That way it will do the cosine calculation and any necessary value wrapping under the hood.

melanieclarke replied:
Actually, I think this is the function I was looking for:
[spherical_offsets_by](https://docs.astropy.org/en/latest/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.spherical_offsets_by)

melanieclarke commented on jwst/cube_build/cube_build_step.py:
```diff
@@ -530,3 +543,55 @@ def read_user_input(self):
+            return None
+
+        for model in self.input_models:
+            file_check = model.meta.filename
+            if file_check in offset_filename:
+                continue
```
Here, we should probably check for and discard any paths in the filenames before matching them (e.g. with os.path.basename).

jemorrison replied:
I have checked that if the association has a filename with a path then the file_check value is just the filename
It is handled in how the association sets model.meta.filename
The issue with the path is if the offset file has the path. I put in the documentation that the offset filename can not have a path in the filename. 
Is this sufficient  or where you thinking of something else ?

melanieclarke replied:
Good to know that model.meta.filename will only contain the basename, though!  Thanks for checking that, I wasn't sure.

melanieclarke commented on jwst/cube_build/ifu_cube.py:
```diff
@@ -1306,13 +1309,47 @@ def setup_ifucube_wcs(self):
+                # pull out ra dec offset if it exists
+                if self.offsets is not None:
+                    filename = input_model.meta.filename
+                    index = self.offsets['filename'].index(filename)
+                    raoffset = self.offsets['raoffset'][index]
+                    decoffset = self.offsets['decoffset'][index]
```
Does the schema require that this is always valid?  That is, is it possible to get to this point with a mismatched file and offset list?

jemorrison replied:
Not sure I understand the question. If the filename is not in the offset file then it gives an error to the user.
If you are asking if the user has not kept the order of the filenames and the raoffset and decoffsets the same then
that is a problem. But that is why I gave the example of how to make the offset file and it seem fairly straight forward that they need to match. 
Is that  what you are asking ?

melanieclarke replied:
Sorry, I wasn't clear here either.  My question was - what happens if the user accidentally provides 4 filenames but 3 RA or Dec offset values?  If it gets to this point in the code in that condition, these lines will throw an IndexError.  

I'm not familiar enough with the ASDF validation to be sure, but it didn't look to me like you were checking the length of the lists anywhere and making sure that they match.  Maybe that check just needs to be added to the check_offset_file function?

jemorrison replied:
Ok I will see if I can add a check for that

jemorrison replied:
I just added it into the code in cube_build as a check.

melanieclarke commented on jwst/cube_build/tests/test_configuration.py:
```diff
@@ -467,3 +539,41 @@ def test_calspec3_config_nirspec_multi(tmp_cwd, nirspec_medium_coverage):
 
     assert cube_pars['1']['par1'] == ['g140m', 'g235m']
     assert cube_pars['1']['par2'] == ['f100lp', 'f170lp']
+
```
These look like good tests for the check_offset_file function, but none of the code that actually applies the offsets is covered.  It looks like there is some synthetic data in some of the other tests that can be run through cube_build successfully - can you write a test or two that applies some example offsets to the synthetic data?

jemorrison replied:
Would it be sufficient to make a regression tests to test applying the offsets ? I am trying to set up a unit test but we do not call the full cube building in the unit tests - we call smaller subsets. The cube_building is done in ifu_cube.py.  If we call ifu_cube.setup_ifucube_wcs - then we could test how the cube coordinate system is shifted. But there is still quite a lot in that routine and it not all the easy form a unit test because it  requires so of the infrastructure to be setup - WCS, WCSINFO, models. Just checking what the real goal is and if a regression test would be sufficient.  I could rewrite setup_ifube_wcs to do less set up  so it is easier to write unit test for it- but not sure that is a good idea either.

melanieclarke replied:
Two goals...

One, most of the code you've added is not currently covered by either regression or unit tests, which makes me nervous.  I think we should add one or the other.

Two, unit test coverage for this module looks pretty minimal in general.  It would be nice to start working on getting that coverage up, so if it's possible to add some unit tests here, it would help with maintainability and robustness in this part of the code base.

So, if possible, it would be nice to add unit tests to cover these changes.  If it is too difficult to manage within the scope of this PR, then we should add a regression test.

It looks to me like there are some tests in test_cube_build_step.py that run the full step, which is why I thought you might be able to build on those and create a test that passes in an offset file.  But I haven't tried, so it may be more difficult than it looks!

jemorrison replied:
I finally have it. I just had to set up more default values to the code so it would run all the way to test ifu_cube.setup_ifucube_wcs

melanieclarke reviewed (changes requested):
I think this needs a little more work still.  Some suggestions and questions below.

Also, sort of unrelated to this PR, but while I was reading cube build docs, I noticed that the examples.rst has some bad formatting (see the top of https://jwst-pipeline.readthedocs.io/en/latest/jwst/cube_build/examples.html).  Can you please clean that up, while we're updating the docs for this step?

jemorrison commented on docs/jwst/cube_build/arguments.rst:
```diff
@@ -122,3 +129,43 @@ A parameter only used for investigating which detector pixels contributed to a c
+   import asdf
+   import astropy.units as u
+   
+   filename = []
+   raoffset = []
+   decoffset = []
+   for i in range(num):
+       filename.append(file[i])
+       raoffset.append(ra_center1[i])
+       decoffset.append(dec_center1[i])
```
I would like to the keep the one with the loop because that is how they will be made in general. I think that is straight forward to me. I will also add your example.

melanieclarke replied:
The reason I asked is that the code in the loop example is not runnable as-is - you'd have to set up the file, ra_center1, dec_center1, and num variables first, and it's not clear how you'd do that.  To keep the example simple, I think it's better to just define the lists.

jemorrison replied:
I wrote in the comments that it is up the user to provide those. I can make that clearer.

jemorrison commented:
Updated the code to use astropy  SkyCoord and the routine to offset Ra and Dec values, spherical_offsets. The main difference is the cos(dec) term.  SkyCoord uses the declination of the coordinate while before we were using the median declination of the data. The benefit of SkyCoord is that it also does the necessary checks on Ra and Dec when adding the offset (i.e Ra says within 0 to 360 and dec stays 90 to -90. )

melanieclarke commented on jwst/cube_build/ifu_cube.py:
```diff
@@ -113,7 +116,8 @@ def __init__(self,
         self.cdelt3_normal = None
         self.rot_angle = None  # rotation angle between Ra-Dec and IFU local instrument plane
-
+        self.median_dec = None
```
I think this is no longer needed

jemorrison replied:
agreed

melanieclarke reviewed (commented):
Thanks for all the updates - I think we're close, now!

There were a couple comments from the previous review that I think GitHub hid from you, if you could please take another look?  In particular:
- I think we don't need to require that the units are specified in the offset file if we require them to always be arcsec
- An invalid offset file should raise an exception, rather than logging an error and continuing.

Also, looking at the SkyCoord changes, it might be nice to put that in a little function to reduce the repeated code that makes a SkyCoord, applies the offset, and pulls the values back out.

A few more suggestions also below.

jemorrison commented on docs/jwst/cube_build/arguments.rst:
```diff
@@ -122,3 +128,83 @@ A parameter only used for investigating which detector pixels contributed to a c
+       filename = []
+       raoffset = []
+       decoffset = []
+       num = len(files)
+       for i in range(num):
+           filename.append(files[i])
+           raoffset.append(ra_offset[i])
+           decoffset.append(dec_offset[i])
```
gone

jemorrison commented on jwst/cube_build/ifu_cube.py:
```diff
@@ -1716,6 +1754,19 @@ def map_miri_pixel_to_sky(self, input_model, this_par1, subtract_background):
+            raoffset = offsets['raoffset'][index]
+            decoffset = offsets['decoffset'][index]
+            log.info("Ra and Dec offset (arc seconds) applied to file :%8.6f, %8.6f,  %s",
+                     raoffset*3600.0,
+                    decoffset*3600.0, filename)
+            raoffset = raoffset* units.deg
+            decoffset = decoffset* units.deg
```
done

jemorrison commented:
@melanieclarke  Ok I think all the requested changes are in the code now. I will look over it again tonight to make there are no typos 
 I will run another regression test now

melanieclarke commented on jwst/cube_build/cube_build_step.py:
```diff
@@ -530,3 +544,64 @@ def read_user_input(self):
+                continue
+            else:
+                af.close()
+                raise Exception('Error in offset file. A file in assocation is not found in offset list list %s', file_check)
```
```suggestion
                raise ValueError('Error in offset file. A file in the association is not found in offset list %s', file_check)
```

melanieclarke commented on jwst/cube_build/cube_build_step.py:
```diff
@@ -530,3 +544,64 @@ def read_user_input(self):
+        len_dec = len(offset_dec)
+        if (len_file != len_ra or len_ra != len_dec or len_file != len_dec):
+            af.close()
+            raise Exception('Offset file error. The offset file does not have the same number of values for filename, offset_ra, offset_dec')
```
```suggestion
            raise ValueError('The offset file does not have the same number of values for filename, raoffset, decoffset.')
```

melanieclarke reviewed (commented):
Updating suggestions for new error messages

melanieclarke reviewed (approved):
Okay, I think all of my comments have been addressed!

I did a quick integrity check on the current branch, by making an offsets file following the example in the documentation, with 1 arcsec offsets in both RA and Dec.  The output cube looks good, with the offset applied at the right magnitude and direction.

In the interest of getting this change into this build, let's go ahead and leave the arcsec units in the schema, and leave it for future work to support other units.

I approve!  And thanks again for your patience in working this through.

karllark reviewed (approved):
Looks good to me.

melanieclarke commented:
One last regtest run to make sure there are no interactions with other merged PRs:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1739/

melanieclarke commented:
No relevant regtest failures.  Merging now!
